### ISSUE TITLE
UrlHelper can't match with any defined routes if missing _method option

### ISSUE THREAD

**Issue by eryw** [cakephp/cakephp#11467]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.5.6
* Platform and Target: PHP built-in webserver

### What you did
In `routes.php` I declared a route like this:
```php
$routes->connect('/login', ['controller' => 'Users', 'action' => 'login'], ['_name' => 'login'])->setMethods(['GET', 'POST']);
```

In the template file, I uses `UrlHelper` like this:
```php
$this->Url->build(['_name' => 'login'])
```

### What happened
I get `Cake\Routing\Exception\MissingRouteException`
and an error
`Warning (2): strtoupper() expects parameter 1 to be string, array given [CORE/src/Routing/Route/Route.php, line 744]`

### What you expected to happen
It should match with the route that I have declared in `routes.php` and return `/login` string.
The error will gone if I'm also add `_method` option like this:
```php
$this->Url->build(['_name' => 'login', '_method' => 'GET'])
```
I think it should not matter what http method will be used the url to be generated is `/login`

**Comment by saeideng** [comment]
do you use custom routeClass ?

**Comment by saeideng** [comment]
`$url['_method'] ` changed in someplace like routeClass

**Comment by markstory** [comment]
How could someone reproduce this? I've not seen `$url['_method']` ever be an array.

**Comment by eryw** [comment]
@saeideng I don't use custom route class. I use built-in `DashedRoute` class.
@markstory `_method` become an array because:
- I call this method `->setMethods(['GET', 'POST'])` in `routes.php`. It will set default method to an array `['GET', 'POST']`.
- I don't provide the `_method` option when calling `Url::build()`. I don't provide it because I use named route, that what I expect is only need to pass the `_name` option.
- At this line, https://github.com/cakephp/cakephp/blob/3.5.6/src/Routing/RouteCollection.php#L316 will use default `_method` which is an array.

**Comment by saeideng** [comment]
confirm 
```php
//routes.php
$routes->connect('/login', ['controller' => 'Users', 'action' => 'login'], ['_name' => 'login'])->setMethods(['GET', 'POST']);

//view
$this->Url->build(['_name' => 'login'])

**Comment by saeideng** [comment]
I can care to this

**Comment by markstory** [comment]
Thanks @saeideng

**Comment by markstory** [comment]
Closing as a pull request is open now.

**Comment by beporter** [comment]
I just stumbled on a  possibly-related fact (which is why I'm mentioning it here first, before opening a separate issue.)

If you create a route using a specific HTTP method then you are not able to reverse route it without explicitly including the `_method`.

```php
// Using `->get()` instead of `->connect()`:
$routes->get('/auth/login', ['controller' => 'Auth', 'action' => 'login']);

// Fails with `Error: A route matching "array ( 'controller' => 'Auth', 'action' => 'login', 'plugin' => NULL, '_ext' => NULL, )" could not be found.`
Router::url(['controller' => 'Auth', 'action' => 'login']);

// Succeeds with explicit `_method` included.
Router::url(['controller' => 'Auth', 'action' => 'login', '_method' => 'GET']);
```

```shell
$ bin/cake routes
+-------------------------------------+-----------------------------------------+----------------------------------------------------------------------------+
| Route name                          | URI template                            | Defaults                                                                   |
+-------------------------------------+-----------------------------------------+----------------------------------------------------------------------------+
| auth:login                          | /auth/login                             | {"_method":"GET","action":"login","controller":"Auth","plugin":null}       |

**snip**
```

Which leads me to the question: Should reverse routes match `[_method => GET]` by default?

**Comment by markstory** [comment]
Router will add the `_method` key from the request context if it exists. Are you generating URLs with no active request?

**Comment by beporter** [comment]
Negative. This is in-browser.

**Comment by beporter** [comment]
I'm looking here https://github.com/cakephp/cakephp/blob/ae7ec5/src/Routing/Route/Route.php#L677 and don't see where the `_method` is **set** if it isn't already present. 😕

**Comment by markstory** [comment]
Hrm I thought we did merge in a `_method`. One option would be to change how methods are matched https://github.com/cakephp/cakephp/blob/ae7ec50e12adbb30d9e7d1187b385bc32b7d31dc/src/Routing/Route/Route.php#L761-L781

to skip method matching if the route is a GET.

**Comment by beporter** [comment]
I had a similar thought, yeah. Provided it doesn't change the operational semantics in a negative way for existing uses. Thinking through it: Would I ever want `Router::url(['controller' => 'Auth', 'action' => 'login' /* no explicit _method! */]);` to _start_ matching a route that it previously didn't because of this change? I can't think of a situation where `GET` isn't an entirely reasonable default. 🤷‍♀️

**Comment by beporter** [comment]
tl;dr: Intuitively I'd expect a reverse lookup without an explicit `[_method => ...]` to match a route that includes _either_ an explicit **or** implicit `[_method => GET]` where `$routes->connect()` routes include an implicit GET and `$routes->get()` are explicit.

(This seems like it's kind of a weird side effect of declared routes that accept "any method" (`Router::connect()`) not explicitly _naming_ all possible methods; instead they use the shortcut of  `_method` being empty or undefined which would logically mean "don't accept any method" even though that's useless.)

**Comment by beporter** [comment]
Honestly it makes more sense to me for reverse routing requests to implicitly include `[_method => GET]` if the programmer doesn't explicitly request a specific method.

```diff
         if (empty($url['_method'])) { 
-            return false; 
+            $url['_method'] = 'GET';
         } 
```
...would probably do the trick and make the intent pretty clear: _"If we don't have an explicit `[_method]` to match, make sure we at least match on `GET`"_.

**Comment by markstory** [comment]
@beporter Any chance you could put a pull request together with that change?
